scripts/upload-daemon/Main.hs

```diff
@@ -0,0 +1,128 @@
+import           System.Metrics.Prometheus.Metric.Counter       as C (Counter,
+                                                                      inc)
+import           System.Metrics.Prometheus.Metric.Gauge         as G (Gauge,
+                                                                      dec, inc)
+import           System.Metrics.Prometheus.MetricId             (addLabel,
+                                                                 fromList)
+import           System.Process                                 (readProcessWithExitCode)
+
+data StatisticsHandlers
+  = StatisticsHandlers
+  { requests :: Counter
+  , success  :: Gauge
+  , failure  :: Gauge
+  , running  :: Gauge
+  }
+
+type UploadTarget = String
+
+-- | First element of a list that is not Nothing
+firstJust :: [Maybe a] -> Maybe a
+firstJust = join . listToMaybe . Prelude.filter isJust
+
+
+-- | Upload target is taken either from first argument or from UPLOAD_TARGET env variable
+getUploadTarget :: IO (Maybe UploadTarget)
+getUploadTarget = firstJust <$> sequence
+  [ listToMaybe <$> getArgs
+  , optional $ getEnv "UPLOAD_TARGET"
+  ]
+
+-- | Get an environment variable with default fallback
+getEnvDefault :: String -> String -> IO String
+getEnvDefault var def = fromMaybe def <$> (optional $ getEnv var)
+
+-- | Receive data from `sock` until EOF or connection closure
+recvAll :: Socket -> IO (Maybe ByteString)
+recvAll sock = do
+  dat <- recv sock 1024
+  case dat of
+    Just str ->
+      if BS.length str == 1024
+      then (\rest -> dat <> rest) <$> recvAll sock
+      else return dat
+    Nothing -> return Nothing
+
+-- | Upload a path to target binary cache
+upload :: UploadTarget -> StatisticsHandlers -> ByteString -> IO ()
+upload target (StatisticsHandlers {..}) path = do
+  G.inc running
+  (code, _, stderr) <- readProcessWithExitCode "nix"
+    [ "copy", "--to", target, (unpack path) ]
+    ""
+  G.dec running
+  if code /= ExitSuccess
+    then G.inc failure >> Prelude.putStrLn stderr
+    else G.inc success >> Prelude.putStr "Uploaded " >> BS.putStrLn path
+
+
+-- | Receive newline-separated paths until EOF, and upload them
+handleConnection :: UploadTarget -> StatisticsHandlers -> (Socket, SockAddr) -> IO ()
+handleConnection uploadTarget handlers@(StatisticsHandlers {..}) (sock, _) = do
+  C.inc requests
+  Just paths <- recvAll sock -- Failure here will fail only the upload thread
+  let pathsList = BS.lines paths
+  Prelude.putStr "Uploading "
+  print pathsList
+  send sock "Queued "
+  send sock $ BS.pack $ show $ Prelude.length pathsList
+  send sock " paths\n"
+  closeSock sock
+  parallel_ $ (upload uploadTarget handlers) <$> pathsList
+
+main :: IO ()
+main = do
+  uploadTarget <- fromJust <$> liftIO getUploadTarget -- Bail out if no upload target is specified
+  port <- getEnvDefault "PORT" "8081"
+  prometheusPort <- getEnvDefault "PROMETHEUS_PORT" "8080"
+  Prelude.putStr "Starting server on localhost:"
```

Comment:
```suggestion
  Prelude.putStr "Starting server on localhost: "
```
(use `fmt`)

Reply:
No, this is indeed intended (example output: `Starting server on localhost:8081 uploading to ssh://t490s-laptop`), but I've replaced it with fmt so that it's more clear :)